### ISSUE TITLE
understanding the current state of dynamic_rnn vs buckets vs scan for seq2seq

### ISSUE THREAD

**Issue by 22csnyder** [tensorflow/tensorflow#2381]
Bear with me as I am not an expert in the practical or theoretical details. 

Perhaps because tensorflow has developed so rapidly, I find it a little difficult to find the recommended procedure for RNNs with varying length, at least from my lay end-user perspective. It is unclear to me on May 15, 2016 what tool within tensorflow is best suited to this task.

It seems that originally RNNs of varying sequence length were handled by "bucketing". This means building a "separate" model for each of several bucket lengths so that inputs could be padded to the same length without wasting too much gpu memory. There is not presently a "simple" example with bucketing, but to those unfamiliar, I would look to the [seq2seq](https://github.com/tensorflow/tensorflow/blob/3c8780451007c27b69f10925d43d1f3501d94106/tensorflow/models/rnn/translate/seq2seq_model.py) model for clarity. Also the [model_with_buckets function](https://github.com/tensorflow/tensorflow/blob/82ff4cd8b0d541ede107d34d8eecc769c91dda11/tensorflow/python/ops/seq2seq.py)

Sometime between March and May a number of (maybe?) different tools were added that could possibly be used to avoid this bucketing trick without wasting memory.

1. [rnn.py/dynamic_rnn](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/rnn.py)
2. [functional_ops](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/functional_ops.py#L256), though within this code it's not clear without investigation if map_fn or tf.scan (!) would be better suited.
3. [control_flow_ops](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/ops/control_flow_ops.py)

In spite of this, the seq2seq model still uses bucketing, and I struggle to find any examples using these newer methods.

Creating detailed documentation is ultimately part of the endgoal, but _for now_ I thought it would be useful to at least make a mention of what the recommended approach is.

Thanks.

[relevant issue 208 ](https://github.com/tensorflow/tensorflow/issues/208)
[relevant reddit with broken link to example](https://www.reddit.com/r/MachineLearning/comments/4dtbks/tensorflow_now_has_an_unofficial_scan_function/)

**Comment by wb14123** [comment]
In my understate, `dynamic_rnn` is not used to solve the problem of variable length, but used to run loops within a tensor. So is control flow ops.

I think the good way to deal with variable length, is to divide one sample into multiple ones, and keep the RNN state between these. So the divided multiple samples is seen as one sample. On the other hand, if one sample is shorter, then padding it with zero or something like that. The implement of Keras gives a good example. This should be a better way than bucketing. Though in TensorFlow, it seems have a lot work to do.

**Comment by wb14123** [comment]
My mistake, it seems that `dynamic_rnn` implemented the way I said. But it doesn't matter with control flow ops, you can do it with python loop.

**Comment by girving** [comment]
@wb14123: A Python loop doesn't support automatic gradient computation, which is the typical reason to use the fancier control flow ops.  I'm going to close this for now, since questions about how to use TensorFlow are better suited to StackOverflow (use the `tensorflow` tag).

**Comment by Lynten** [comment]
@girving I'm somewhat confused.

In [static_rnn](https://github.com/tensorflow/tensorflow/blob/1752d9c8fac5f6cf85a41e77d92e2743adbfc446/tensorflow/python/ops/rnn.py#L1312), it does use Python loop to compute rnn outputs and final state.
```
    for time, input_ in enumerate(inputs):
      if time > 0:
        varscope.reuse_variables()
      # pylint: disable=cell-var-from-loop
      call_cell = lambda: cell(input_, state)
      # pylint: enable=cell-var-from-loop
      if sequence_length is not None:
        (output, state) = _rnn_step(
            time=time,
            sequence_length=sequence_length,
            min_sequence_length=min_sequence_length,
            max_sequence_length=max_sequence_length,
            zero_output=zero_output,
            state=state,
            call_cell=call_cell,
            state_size=cell.state_size)
      else:
        (output, state) = call_cell()

      outputs.append(output)
```

You said "a Python loop doesn't support automatic gradient computation", then how does the static_rnn work?
